service/src/test/java/bio/terra/workspace/app/configuration/external/controller/FolderApiControllerTest.java

```diff
@@ -0,0 +1,310 @@
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import bio.terra.workspace.common.BaseUnitTest;
+import bio.terra.workspace.generated.model.ApiCreateFolderRequestBody;
+import bio.terra.workspace.generated.model.ApiFolderDescription;
+import bio.terra.workspace.generated.model.ApiFolderDescriptionsList;
+import bio.terra.workspace.generated.model.ApiUpdateFolderRequestBody;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.Optional;
+import java.util.UUID;
+import javax.annotation.Nullable;
+import org.apache.http.HttpStatus;
+import org.broadinstitute.dsde.workbench.client.sam.model.UserStatusInfo;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+public class FolderApiControllerTest extends BaseUnitTest {
+  private static final AuthenticatedUserRequest USER_REQUEST =
+      new AuthenticatedUserRequest(
+          "fake@email.com", "subjectId123456", Optional.of("ThisIsNotARealBearerToken"));
+
+  @Autowired MockMvc mockMvc;
+  @Autowired ObjectMapper objectMapper;
+
+  @MockBean SamService mockSamService;
+
+  @BeforeEach
+  public void setUp() throws InterruptedException {
+    // Needed for workspace creation as logging is triggered when a workspace is created in
+    // `WorkspaceActivityLogHook` where we extract the user request information and log it to
+    // change agents.
+    when(mockSamService.getUserStatusInfo(any()))
+        .thenReturn(
+            new UserStatusInfo()
+                .userEmail(USER_REQUEST.getEmail())
+                .userSubjectId(USER_REQUEST.getSubjectId()));
+  }
+
+  @Test
+  public void createFolder_topLevelFolderCreated() throws Exception {
+    UUID workspaceId = createDefaultWorkspace(mockMvc, objectMapper).getId();
+    var displayName = "foo";
+    var description = String.format("This is folder %s", displayName);
+
+    ApiFolderDescription folderDescription =
+        createFolder(workspaceId, displayName, description, null);
+
+    assertEquals(displayName, folderDescription.getDisplayName());
+    assertEquals(description, folderDescription.getDescription());
+    assertNotNull(folderDescription.getId());
+    assertNull(folderDescription.getParentFolderId());
+  }
+
+  @Test
+  public void createFolder_parentFolderDoesNotExist_throws404() throws Exception {
```

Comment:
Can you have `_throws403` and `_throws404` tests for the various endpoints that throw them?